docs/dg/dev/guidelines/performance-guidelines/custom-code-performance-guidelines.md

```diff
@@ -0,0 +1,805 @@
+```php
+// Bad: Caching data accessed once per week
+$this->storageClient->setex("admin:report:monthly", 86400, $hugeReport);
+```
+
+## Avoid heavy processing in web requests
+
+Web requests must respond quickly—typically under 200-500ms (in some cases - up to 1s). Heavy processing operations should execute in background jobs to maintain user experience and system stability.
+
+### Operations that belong in background jobs
+
+**1. File processing:**
+
+```php
+// ❌ Bad: Process file upload during request
+public function uploadProductImages(array $files): void
+{
+    foreach ($files as $file) {
+        $image = $this->imageProcessor->resize($file, [800, 600]);
+        $thumbnail = $this->imageProcessor->resize($file, [200, 200]);
+        $this->imageProcessor->optimize($image);
+        $this->imageProcessor->optimize($thumbnail);
+        $this->saveToStorage($image, $thumbnail);
+    }
+}
+
+// ✅ Good: Queue for background processing
+public function uploadProductImages(array $files): void
+{
+    foreach ($files as $file) {
+        $this->queueClient->sendMessage('image-processing', [
+            'file_path' => $this->saveTemporaryFile($file),
+            'product_id' => $this->getProductId(),
+        ]);
+    }
+}
+```
+
+**2. PDF generation:**
+
+```php
+// ❌ Bad: Generate PDF synchronously
+public function generateInvoicePdf(int $orderId): Response
+{
+    $pdf = $this->pdfGenerator->create($orderId); // Takes 5-10 seconds
+    return new Response($pdf, 200, ['Content-Type' => 'application/pdf']);
+}
+
+// ✅ Good: Generate asynchronously, notify when ready
+public function requestInvoicePdf(int $orderId): Response
+{
+    $this->queueClient->sendMessage('pdf-generation', ['order_id' => $orderId]);
+
+    return new JsonResponse([
+        'message' => 'PDF generation started',
+        'status_url' => "/api/invoice/{$orderId}/status",
+    ]);
+}
+```
+
+**3. Email sending:**
+
+```php
+// ❌ Bad: Send emails synchronously
+public function placeOrder(QuoteTransfer $quote): OrderTransfer
+{
+    $order = $this->createOrder($quote);
+
+    // Blocks order placement for 2-5 seconds
+    $this->emailService->sendOrderConfirmation($order);
+    $this->emailService->sendInternalNotification($order);
+
+    return $order;
+}
+
+// ✅ Good: Queue emails for background sending
+public function placeOrder(QuoteTransfer $quote): OrderTransfer
+{
+    $order = $this->createOrder($quote);
+
+    // Queue email events
+    $this->eventFacade->trigger('Order.placed', $order);
+
+    return $order;
+}
+```
+
+**4. Data imports and synchronization:**
+
+```php
+// ❌ Bad: Import during admin action
+public function importProductsAction(): Response
+{
+    $products = $this->fetchFromExternalSystem(); // 30+ seconds
+    $this->importProducts($products);
+
+    return new Response('Import completed');
+}
+
+// ✅ Good: Trigger background import job
+public function importProductsAction(): Response
+{
+    $jobId = $this->scheduleImportJob('products');
```

Comment:
I know that we do not have this OOTB, I implemented it for a project POC. We can bring it to Core if you think this is a good idea and jenkins will be our scheduler in the future.